bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryDirectDataWriterHelper.java

```diff
@@ -184,19 +194,81 @@ public void addRow(ByteString message) throws IOException {
     appendRequestRowCount++;
   }
 
+  /**
+   * Thorws an exception if there is an error in any of the responses received thus far. Since
+   * responses arrive in order, we proceed to check the next response only after the previous
+   * response has arrived.
+   *
+   * @throws BigQueryConnectorException If there was an invalid response
+   */
+  private void checkForFailedResponse(boolean waitForResponse) {
+    while (nextIndexCheckResponse < appendRowsFutures.size()) {
+      ApiFuture<AppendRowsResponse> validatedAppendRowsFuture =
+          appendRowsFutures.get(nextIndexCheckResponse);
+      if (waitForResponse || validatedAppendRowsFuture.isDone()) {
+        boolean succeeded = false;
+        try {
+          AppendRowsResponse response = validatedAppendRowsFuture.get();
+          // If we got here, the response was successful
+          succeeded = true;
+          nextIndexCheckResponse++;
+        } catch (ExecutionException e) {
+          if (e.getCause().getClass() == OffsetAlreadyExists.class) {
+            // Under heavy load the WriteAPI client library may resend a request due to failed
+            // connection. The OffsetAlreadyExists exception indicates a request was previously
+            // sent, hence it is safe to ignore it.
+            logger.warn("Ignoring OffsetAlreadyExists exception: {}", e.getCause().getMessage());
+            succeeded = true;
+            nextIndexCheckResponse++;
+          } else {
+            logger.error(
+                "Write-stream {} with name {} exception while retrieving AppendRowsResponse",
+                partitionId,
+                writeStreamName);
+            throw new BigQueryConnectorException(
+                "Exception while retrieving AppendRowsResponse", e);
+          }
+        } catch (InterruptedException e) {
+          logger.error(
+              "Write-stream {} with name {} exception while retrieving AppendRowsResponse",
```

Comment:
*interrupted while retrieving?
It's preferable not to have same message point to different possible causes.

Reply:
Done.